From: Frediano Ziglio <frediano.ziglio@citrix.com>
Date: Fri, 13 Mar 2026 17:00:15 +0100
Subject: XSI-2150: Avoid deadlocks on NFS spinlocks

Due to some backport trying to fix a possible deadlock (see CA-399062
for details) some patches which introduced many more deadlock were
merged to XS8.
Revert these patches and fix the original potential deadlock too.

These deadlock were generated by the same lock taken by no-IRQ context
and nested lock attempt on IRQ context as IRQ were not disabled on
first lock. Calling spin_lock_bh instead of spin_lock prevent IRQ
code nested to no-IRQ code.

The removal of "transport_lock" in "xs_tcp_state_change" is enabled by
commit 13f6175 ("XSI-2150: Remove added chunk causing deadlock",
specifically "SUNRPC: Move reset of TCP state variables into the
reconnect code"). A review of "xs_tcp_state_change" shows that similar
accesses through the "transport" pointer (notably for "TCP_CLOSE_WAIT"
and "TCP_CLOSE") are already performed without this lock. State bit
updates are atomic, and the work queue is adequately protected by
other synchronization mechanisms, including additional spinlocks and
CPU-local variables.

diff --git a/net/sunrpc/xprt.c b/net/sunrpc/xprt.c
index 8c1aaa0836ea..5da6612f8def 100644
--- a/net/sunrpc/xprt.c
+++ b/net/sunrpc/xprt.c
@@ -765,18 +765,18 @@ xprt_init_autodisconnect(struct timer_list *t)
 {
 	struct rpc_xprt *xprt = from_timer(xprt, t, timer);
 
-	spin_lock(&xprt->transport_lock);
+	spin_lock_bh(&xprt->transport_lock);
 	if (!list_empty(&xprt->recv))
 		goto out_abort;
 	/* Reset xprt->last_used to avoid connect/autodisconnect cycling */
 	xprt->last_used = jiffies;
 	if (test_and_set_bit(XPRT_LOCKED, &xprt->state))
 		goto out_abort;
-	spin_unlock(&xprt->transport_lock);
+	spin_unlock_bh(&xprt->transport_lock);
 	queue_work(xprtiod_workqueue, &xprt->task_cleanup);
 	return;
 out_abort:
-	spin_unlock(&xprt->transport_lock);
+	spin_unlock_bh(&xprt->transport_lock);
 }
 
 bool xprt_lock_connect(struct rpc_xprt *xprt,
diff --git a/net/sunrpc/xprtsock.c b/net/sunrpc/xprtsock.c
index e542768f5d3d..834e2b8be72b 100644
--- a/net/sunrpc/xprtsock.c
+++ b/net/sunrpc/xprtsock.c
@@ -1611,7 +1611,6 @@ static void xs_tcp_state_change(struct sock *sk)
 	trace_rpc_socket_state_change(xprt, sk->sk_socket);
 	switch (sk->sk_state) {
 	case TCP_ESTABLISHED:
-		spin_lock(&xprt->transport_lock);
 		if (!xprt_test_and_set_connected(xprt)) {
 			xprt->connect_cookie++;
 			clear_bit(XPRT_SOCK_CONNECTING, &transport->sock_state);
@@ -1619,7 +1618,6 @@ static void xs_tcp_state_change(struct sock *sk)
 
 			xs_run_error_worker(transport, XPRT_SOCK_WAKE_PENDING);
 		}
-		spin_unlock(&xprt->transport_lock);
 		break;
 	case TCP_FIN_WAIT1:
 		/* The client initiated a shutdown of the socket */
